test/compile.jl

```diff
@@ -204,16 +204,19 @@ try
         @test stringmime("text/plain", Base.Docs.doc(Foo.Bar.bar)) == "bar function\n"
 
         modules, deps, required_modules = Base.parse_cache_header(cachefile)
+        discard_module = mod_fl_mt -> (mod_fl_mt[2], mod_fl_mt[3])
         @test modules == Dict(Foo_module => Base.module_uuid(Foo))
-        @test map(x -> x[1],  sort(deps)) == [Foo_file, joinpath(dir, "bar.jl"), joinpath(dir, "foo.jl")]
+        @test map(x -> x[1],  sort(discard_module.(deps))) == [Foo_file, joinpath(dir, "bar.jl"), joinpath(dir, "foo.jl")]
+        srctxt = Base.read_dependency_src(cachefile, Foo_file)
+        @test !isempty(srctxt) && srctxt == read(Foo_file, String)
```

Comment:
Can we add a test for a manually included dependency. If I understand one could end up parsing the list of dependent files and end up with an `("__external__", "external.so", mtime)` and `Base.read_dependency_src(cachefile, "external.so") should fail.

Reply:
Right, but IMO that's fine. This is an internal method and consumers can make sure they don't try to read `"__external__"` dependencies. (Reading the src text is not something that's done unless it is requested.) Revise doesn't intend to track external dependencies, because it's hard to know what you're supposed to do about them.

But if you know of packages that are using that, I would be happy to go look at how they're using external dependencies; that could end up changing my mind.

Reply:
Everything that would lead to a stale cachefile should probably prompt Revise to to reload that file :).
In my use cases I have stateful information from that external dependency (if it exists or not, API version, ...)

Reply:
Would love a link to your use cases, if they are in public repos.

It would be fine to trigger `revise(module)` when an external dependency updates, but it seems like it can't be robust: what if the external dependency is a single file in another package? How is Revise supposed to know whether it has to revise *that* package first? (In some cases it wouldn't but in others it would be necessary.)

In either case, I don't think Revise will need access to the source text of external dependencies. If, as you say, it's an `.so` file, Revise can't do anything with it anyway. In fact it occurs to me that we probably shouldn't store external dependencies in the src-text cache; it would be crazy to cache, say, `libkde.so` in a `*.ji` file :smile:.

I will change the format so that each source-text file has its name at the beginning, so that one doesn't have to "count" from the list of dependencies while skipping over `#__external__` dependencies.

Reply:
I was wondering if that might happen ;) yeah that would be crazy.